### PR TITLE
Order impl

### PR DESCRIPTION
**Pull request by kxnxc**

### CODE REVIEW

File: src/main/java/com/example/onlinebookstore/controller/OrderItemController.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+                                                                   Long orderId) {
+        return orderItemService.getAllOrderItemsByOrderId(orderId);
+    }
+
+    @Operation(summary = "Get specific order item by id",
+            description = "Get specific order item by id")
+    @PreAuthorize("hasAnyAuthority('USER', 'ADMIN')")
+    @GetMapping("/{itemId}")
+    public OrderItemResponseDto getSpecificOrderItem(@PathVariable Long itemId) {
+        return orderItemService.getOrderItemById(itemId);
```
Comment (Rommelua) [inline src/main/java/com/example/onlinebookstore/controller/OrderItemController.java:36]: move this to OrderController and delete class OrderItemController

File: src/main/java/com/example/onlinebookstore/mapper/OrderItemMapper.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    @Mapping(target = "bookId", source = "book.id")
+    OrderItemResponseDto toDto(OrderItem orderItem);
+
+    @Named("getOrderItemDtos")
+    default Set<OrderItemResponseDto> getOrderItemDtos(Set<OrderItem> orderItems) {
+        Set<OrderItemResponseDto> orderItemResponseDtos = new HashSet<>();
+        for (OrderItem orderItem: orderItems) {
+            orderItemResponseDtos.add(toDto(orderItem));
+        }
+        return orderItemResponseDtos;
```
Comment (Rommelua) [inline src/main/java/com/example/onlinebookstore/mapper/OrderItemMapper.java:23]: use Stream API

File: src/main/java/com/example/onlinebookstore/model/Order.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import java.time.LocalDateTime;
+import java.util.HashSet;
+import java.util.Set;
+import lombok.Data;
+import org.hibernate.annotations.JdbcTypeCode;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+import org.hibernate.type.SqlTypes;
+
+@Data
```
Comment (Rommelua) [inline src/main/java/com/example/onlinebookstore/model/Order.java:25]: better do not use `@Data` with etntity class

File: src/main/java/com/example/onlinebookstore/model/Order.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    @Enumerated(EnumType.STRING)
+    @JdbcTypeCode(SqlTypes.VARCHAR)
+    private Status status;
+    @Column(nullable = false)
+    private BigDecimal total;
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
+    @Column(nullable = false)
+    private String shippingAddress;
+    @OneToMany(mappedBy = "order", fetch = FetchType.EAGER)
```
Comment (Rommelua) [inline src/main/java/com/example/onlinebookstore/model/Order.java:46]: do not use FetchType.EAGER

File: src/main/java/com/example/onlinebookstore/repository/shoppingcart/ShoppingCartRepository.java [hunk 5]
```diff
@@ -1,10 +1,23 @@
...
-    Optional<ShoppingCart> findByUser(User user);
+    @EntityGraph(attributePaths = "cartItems")
+    Optional<ShoppingCart> findByUser_Id(Long id);
+
+    @Query(
+            """
+            SELECT sc FROM ShoppingCart sc 
+            LEFT JOIN FETCH sc.cartItems c WHERE sc.user.id = :id
+            """
+    )
```
Comment (Rommelua) [inline src/main/java/com/example/onlinebookstore/repository/shoppingcart/ShoppingCartRepository.java:20]: use @EntityGraph

File: src/main/java/com/example/onlinebookstore/service/impl/OrderItemServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    private final OrderItemMapper orderItemMapper;
+
+    private final Consumer<OrderItem> setOrder = oi -> oi.setOrder(null);
+
+    @Override
+    public List<OrderItemResponseDto> getAllOrderItemsByOrderId(Long orderId) {
+        Set<OrderItem> orderItems = orderItemRepository.findAllByOrderId(orderId);
+        return orderItems
+                .stream()
+                .peek(oi -> oi.setOrder(null))
```
Comment (Rommelua) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderItemServiceImpl.java:29]: what is it for? remove peek method call

File: src/main/java/com/example/onlinebookstore/service/impl/OrderItemServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@RequiredArgsConstructor
+@Service
+public class OrderItemServiceImpl implements OrderItemService {
+    private final OrderItemRepository orderItemRepository;
+    private final OrderItemMapper orderItemMapper;
+
+    private final Consumer<OrderItem> setOrder = oi -> oi.setOrder(null);
```
Comment (Rommelua) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderItemServiceImpl.java:22]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/service/impl/OrderItemServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+    @Override
+    public OrderItemResponseDto getOrderItemById(Long itemId) {
+        Optional<OrderItem> optionalOrderItem = orderItemRepository.findById(itemId);
+        optionalOrderItem.ifPresent(setOrder);
+        return optionalOrderItem
+                .map(orderItemMapper::toDto)
+                .orElseThrow(() ->
+                        new EntityNotFoundException("Can't find order item by id "
+                                + itemId));
```
Comment (Rommelua) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderItemServiceImpl.java:42]:
```suggestion
        return orderItemRepository.findById(itemId)
                .map(orderItemMapper::toDto)
                .orElseThrow(() ->
                        new EntityNotFoundException("Can't find order item by id "
                                + itemId));
```

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        List<Order> allByUser = orderRepository.getAllByUser(user);
+        List<Order> orders = allByUser
+                .stream()
+                .peek(o -> o.getOrderItems()
+                        .forEach(ci -> ci.setOrder(null)))
+                .toList();
+        for (Order order: orders) {
+            orderResponseDtos.add(orderMapper.toDto(order));
+        }
+        return orderResponseDtos;
```
Comment (Rommelua) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java:70]:
```suggestion
        return orderRepository.getAllByUser(user)
            .map(orderMapper::toDto)
            .toList();
```

File: src/main/java/com/example/onlinebookstore/dto/order/OrderStatusRequestDto.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package com.example.onlinebookstore.dto.order;
+
+import jakarta.validation.constraints.NotEmpty;
+import lombok.Data;
+
+@Data
+public class OrderStatusRequestDto {
+    @NotEmpty
+    private String status;
```
Comment (Rommelua) [inline src/main/java/com/example/onlinebookstore/dto/order/OrderStatusRequestDto.java:9]:
```suggestion
    private Order.Status status;
```

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+            }
+        }
+        if (updatedStatus == null) {
+            throw new RuntimeException("Can't find status with name "
+                    + requestDto.getStatus());
+        }
+        order.setStatus(updatedStatus);
+        orderRepository.save(order);
+        order.setOrderItems(orderItemRepository.findAllByOrder(order));
+        return orderMapper.toDto(order);
```
Comment (Rommelua) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java:95]:
```suggestion
        Order order = orderRepository
                .findById(id)
                .orElseThrow(() ->
                        new EntityNotFoundException("Can't find order with id "
                                + id));
        order.setStatus(requestDto.getStatus());
        return orderMapper.toDto(order);
```

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    private BigDecimal getTotalPrice(ShoppingCart shoppingCart) {
+        BigDecimal totalPrice = BigDecimal.ZERO;
+        for (CartItem cartItem: shoppingCart.getCartItems()) {
+            BigDecimal bookPrice = cartItem.getBook().getPrice();
+            BigDecimal quantity = BigDecimal.valueOf(cartItem.getQuantity());
```
Comment (Rommelua) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java:102]: use Steram API

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+            totalPrice = totalPrice.add(bookPrice.multiply(quantity));
+        }
+        return totalPrice;
+    }
+
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    public Set<OrderItem> createOrderItems(ShoppingCart shoppingCart,
```
Comment (Rommelua) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java:109]:
```suggestion
    private Set<OrderItem> createOrderItems(ShoppingCart shoppingCart,
```

File: src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+                                            Order order) {
+        Order mockOrder = new Order();
+        mockOrder.setId(order.getId());
+        Set<CartItem> cartItems = shoppingCart.getCartItems();
+        Set<OrderItem> orderItems = new HashSet<>();
+        for (CartItem cartItem: cartItems) {
```
Comment (Rommelua) [inline src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java:115]: use Steram API

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Review by koval-kn — APPROVED** [review]

**Pull request merged**